packages/components/src/public/files/TemplateDownloadButton.tsx

```diff
@@ -1,35 +1,143 @@
-import React, { FC, memo } from 'react';
+import React, { FC, memo, useCallback, useEffect, useMemo, useState } from 'react';
 
 import { PermissionTypes } from '@labkey/api';
 
 import { User } from '../../internal/components/base/models/User';
 import { RequiresPermission } from '../../internal/components/base/Permissions';
+import { DropdownButton, MenuItem } from '../../internal/dropdowns';
+import { ImportTemplate } from '../QueryInfo';
+import { LoadingSpinner } from '../../internal/components/base/LoadingSpinner';
+import { SchemaQuery } from '../SchemaQuery';
+import { useAppContext } from '../../internal/AppContext';
+import { downloadAttachment } from '../../internal/util/utils';
 
 interface Props {
     className?: string;
-    onClick?: () => void;
-    templateUrl?: string;
+    defaultTemplateUrl?: string;
+    isGridRenderer?: boolean;
+    onDownloadDefault?: () => void;
+    schemaQuery?: SchemaQuery;
     text?: string;
     user?: User;
+    dropDownClassName?: string;
 }
 
 export const TemplateDownloadButton: FC<Props> = memo(props => {
-    const { className, onClick, templateUrl, text = 'Template', user } = props;
+    const {
+        schemaQuery,
+        isGridRenderer,
+        className,
+        onDownloadDefault,
+        defaultTemplateUrl,
+        text = 'Template',
+        user,
+        dropDownClassName,
+    } = props;
+    const [customTemplates, setCustomTemplates] = useState<ImportTemplate[]>();
+    const [loadingTemplates, setLoadingTemplates] = useState<boolean>(false);
+    const { api } = useAppContext();
 
-    if (!onClick && !templateUrl?.length) return null;
+    useEffect(() => {
+        if (!schemaQuery || isGridRenderer || customTemplates) return;
+
+        (async () => {
+            await loadTemplates();
+        })();
+    }, [schemaQuery, isGridRenderer]);
```

Comment:
should `customTemplates` be added to the dependency array here? or is this something that should only be called on mount and the dependency array should be empty?